File: client/py/synnax/synnax.py
#  Copyright 2026 Synnax Labs, Inc.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.

import warnings

from alamos import NOOP, Instrumentation
from freighter import URL

from synnax import (
    access,
    arc,
    auth,
    channel,
    control,
    device,
    framer,
    group,
    ontology,
    rack,
    ranger,
    signals,
    status,
    task,
    user,
)
from synnax.config import try_load_options_if_none_provided
from synnax.options import Options
from synnax.telem import TimeSpan
from synnax.transport import Transport


class Synnax(framer.Client):
    """Client to perform operations against a Synnax cluster.

    If using the python client for data analysis/personal use, the easiest way to
    connect is to use the `synnax login` command, which will prompt and securely
    store your credentials. The client can then be initialized without parameters. When
    using the client in a production environment, it's best to provide the connection
    parameter as arguments loaded from a configuration or environment variable.

    After running the synnax login command::
        client = Synnax()

    Without running the synnax login command::
        client = Synnax(
            host="synnax.example.com",
            port=9090,
            username="synnax",
            password="seldon",
            secure=True
        )
    """

    channels: channel.Client
    access: access.Client
    users: user.Client
    ranges: ranger.Client
    control: control.Client
    signals: signals.Registry
    racks: rack.Client
    devices: device.Client
    tasks: task.Client
    ontology: ontology.Client
    statuses: status.Client
    arcs: arc.Client
    groups: group.Client

    _transport: Transport

    def __init__(
        self,
        host: str = "",
        port: int = 0,
        username: str = "",
        password: str = "",
        secure: bool = False,
        open_timeout: TimeSpan = TimeSpan.SECOND * 5,
        read_timeout: TimeSpan = TimeSpan.SECOND * 5,
        keep_alive: TimeSpan = TimeSpan.SECOND * 30,
        max_retries: int = 3,
        instrumentation: Instrumentation = NOOP,
        cache_channels: bool = True,
    ):
        """Creates a new client. Connection parameters can be provided as arguments, or,
        if none are provided, the client will attempt to load them from the Synnax
        configuration file (~/.synnax/config.json) as well as credentials stored in the
        operating system's keychain.

        If using the client for data analysis/personal use, the easiest way to connect
        is to use the `synnax login` command, which will prompt and securely store your
        credentials. The client can then be initialized without parameters.

        :param host: Hostname of a node in the Synnax cluster.
        :param port: Port of the node.
        :param username: Username to authenticate with.
        :param password: Password to authenticate with.
        :param secure: Whether to use TLS when connecting to the cluster.
        """
        opts = try_load_options_if_none_provided(host, port, username, password, secure)
        self._transport = _configure_transport(
            opts=opts,
            open_timeout=open_timeout,
            read_timeout=read_timeout,
            keep_alive=keep_alive,
            max_retries=max_retries,
            instrumentation=instrumentation,
        )
        self.auth = auth.Client(
            transport=self._transport.unary,
            username=opts.username,
            password=opts.password,
        )
        self.auth.authenticate()
        self._transport.use(self.auth.middleware())
        self._transport.use_async(self.auth.async_middleware())

        ch_retriever = channel.ClusterRetriever(self._transport.unary, instrumentation)
        if cache_channels:
            ch_retriever = channel.CacheRetriever(ch_retriever, instrumentation)
        deleter = framer.Deleter(self._transport.unary, instrumentation)
        ch_creator = channel.Writer(
            self._transport.unary,
            instrumentation,
            ch_retriever if cache_channels else None,
        )
        super().__init__(
            stream_client=self._transport.stream,
            async_client=self._transport.stream_async,
            unary_client=self._transport.unary,
            retriever=ch_retriever,
            deleter=deleter,
            instrumentation=instrumentation,
        )
        self.groups = group.Client(self._transport.unary)
        self.ontology = ontology.Client(client=self._transport.unary)
        self.channels = channel.Client(self, ch_retriever, ch_creator)
        range_retriever = ranger.Retriever(self._transport.unary, instrumentation)
        range_creator = ranger.Writer(self._transport.unary, instrumentation)
        self.signals = signals.Registry(frame_client=self, channels=ch_retriever)
        self.racks = rack.Client(client=self._transport.unary)
        self.devices = device.Client(client=self._transport.unary)
        self.tasks = task.Client(
            client=self._transport.unary,
            frame_client=self,
            rack_client=self.racks,
            device_client=self.devices,
        )
        self.ranges = ranger.Client(
            unary_client=self._transport.unary,
            frame_client=self,
            channel_retriever=ch_retriever,
            writer=range_creator,
            retriever=range_retriever,
            signals=self.signals,
            ontology=self.ontology,
            tasks=self.tasks,
        )
        self.control = control.Client(self, ch_retriever)
        self.users = user.Client(self._transport.unary)
        self.statuses = status.Client(self._transport.unary)
        self.arcs = arc.Client(self._transport.unary)
        self.access = access.Client(self._transport.unary)

    @property
    def hardware(self) -> "Synnax":
        """Deprecated: Use client.devices, client.tasks, client.racks directly."""
        warnings.warn(
            "client.hardware is deprecated and will be removed in a future version. "
            "Use client.devices, client.tasks, client.racks directly instead.",
            FutureWarning,
            stacklevel=2,
        )
        return self

    def close(self):
        """Shuts down the client and closes all connections. All open iterators or
        writers must be closed before calling this method.
        """
        # No-op for now, we'll definitely add cleanup logic in the future, so it's
        # good to have this API defined.
        ...


def _configure_transport(
    opts: Options,
    open_timeout: TimeSpan,
    read_timeout: TimeSpan,
    keep_alive: TimeSpan,
    max_retries: int,
    instrumentation: Instrumentation = NOOP,
) -> Transport:
    t = Transport(
        instrumentation=instrumentation,
        url=URL(host=opts.host, port=opts.port),
        secure=opts.secure,
        open_timeout=open_timeout,
        read_timeout=read_timeout,
        keep_alive=keep_alive,
        max_retries=max_retries,
    )
    return t

File: client/clib/synnax.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include <exception>
#include <memory>
#include <string>

#include "client/clib/internal.h"
#include "client/clib/synnax.h"

using namespace synnax::clib;

int32_t synnax_client_open(
    const char *host,
    const uint16_t port,
    const char *username,
    const char *password,
    const int32_t secure,
    const uint32_t max_retries,
    const int64_t clock_skew_threshold,
    SynnaxClient **out_client,
    SynnaxError *err
) {
    clear_err(err);
    if (out_client == nullptr) {
        set_err(err, CODE_INTERNAL, "sy.validation", "null out_client");
        return CODE_INTERNAL;
    }
    try {
        synnax::Config c;
        c.host = str_or(host, "localhost");
        if (port != 0) c.port = port;
        c.username = str_or(username, "synnax");
        c.password = str_or(password, "seldon");
        c.secure = secure != 0;
        if (max_retries != 0) c.max_retries = max_retries;
        if (clock_skew_threshold > 0)
            c.clock_skew_threshold = x::telem::TimeSpan(clock_skew_threshold);

        auto wrapper = std::make_unique<SynnaxClient>(c);
        const auto state = wrapper->client.connectivity->check();
        if (state.status != synnax::connection::Status::CONNECTED) {
            const bool has_err = !state.error.ok();
            set_err(
                err,
                CODE_ERROR,
                has_err ? state.error.type : std::string("sy.connection"),
                has_err ? state.error.data : state.message
            );
            return CODE_ERROR;
        }
        *out_client = wrapper.release();
        return CODE_OK;
    } catch (const std::exception &e) {
        set_err(err, CODE_INTERNAL, "sy.internal", e.what());
        return CODE_INTERNAL;
    } catch (...) {
        set_err(err, CODE_INTERNAL, "sy.internal", "unknown exception");
        return CODE_INTERNAL;
    }
}

void synnax_client_close(SynnaxClient *client) {
    delete client;
}

const char *synnax_client_version(void) {
    return SYNNAX_VERSION;
}
